### PR TITLE
Add infrastructure to compute errors against observations

### LINKED ISSUE #407: O1.5.3e (coupler) Leaderboard for AMIP
We need to add better metrics (e.g. cloud area fraction, MSE, lapse rate, ...) to time aggregate diagnostics and present differences from observations along the lines of the metrics in the OKR document, compared to the observations. This will be applied to our long runs.

```[tasklist]
### Infrastructure
- [x] [Add infrastructure to compute errors against observations ](https://github.com/CliMA/ClimaCoupler.jl/pull/625)
- [x] date splitting
- [x] Add the seasonal plot summary
- [ ] https://github.com/CliMA/ClimaCoupler.jl/pull/821
- [x] Add RMSE compared to other models
- [ ] https://github.com/CliMA/ClimaCoupler.jl/pull/831
```

```[tasklist]
### Small tasks
- [x] https://github.com/CliMA/ClimaAtmos.jl/issues/2578
- [ ] https://github.com/CliMA/ClimaCoupler.jl/issues/781
- [ ] https://github.com/CliMA/ClimaArtifacts/pull/18
- [ ] https://github.com/CliMA/ClimaCoupler.jl/pull/846
```

### PR DESCRIPTION
**Pull request by Sbozzolo**
This PR introduces a new module, `Leaderboard`, currently living in the AMIP experiment.

`Leaderboard` uses `NCDatasets` and `ClimaAnalysis` to compute error maps between observations and simulations.

`Leaderboard` defines two new structs `ObsDataSource` and `SimDataSource`. 
`ObsDataSource` describes some observational data (contained a NetCDF file) by specifying the name of the variable and other attributes. `SimDataSource` does the same for simulated (atmos) data.

The `rmse` error takes these two data structures and a target date, resamples the observational data to the simulated output grid (with a 0th order interpolation scheme, ie nearest neighbors). Then, it computes
```julia
bias = (sim_data - obs_data)
```
This variable is neatly packaged in a `ClimaAnalysis.OutputVar` so that it can be directly plotted.

`Leaderboard` also provides an numerical integration routine on the sphere. With `integrate_on_sphere` the RMSE can be integrated over the sphere to condense it to a single number (a 1/4pi might be needed to normalize the integration volume). These values are saved in the attributes of the `OutputVar` produced by `bias`.

Closes #407

### CODE REVIEW

**Comment by Sbozzolo** [comment]
image

File: experiments/AMIP/user_io/leaderboard/compare_with_obs.jl [hunk 1]
```diff
@@ -0,0 +1,19 @@
+##########
+# Precip #
+##########
+
+precip_obs = ObsDataSource(;
+    path = "OBS/gpcp.precip.mon.mean.197901-202305.nc",
+    var_name = "precip",
+    # 1 mm/day -> 1 m/s
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/compare_with_obs.jl:8]:
```suggestion
    # 1 mm/day -> - 1 kg/m/s2
```
(to be consistent with the latest version of climaatmos)

File: experiments/AMIP/user_io/leaderboard/compare_with_obs.jl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    unit_conversion_factor = Float32(1e-3 / 86400),
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/compare_with_obs.jl:9]:
```suggestion
    unit_conversion_factor = Float32(-1/ 86400),
```
(If we have access to parameters here, it's better to use Float32(-1e-3 * ρ_liq / 86400 ) ), where ρ_liq is defined [here](https://github.com/CliMA/ClimaCoupler.jl/blob/b57d37457fb0a12e577f7633596cabea8473dbe2/experiments/AMIP/components/land/bucket_utils.jl#L56)

File: experiments/AMIP/user_io/leaderboard/data_sources.jl [hunk 3]
```diff
@@ -0,0 +1,111 @@
+"""
+
+    A struct to describe some observation data that we want to compare against.
+
+"""
+struct ObsDataSource
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/data_sources.jl:6]: Just note that we will compare with data that has a z dimension as well, but we can add that when we are actually doing it.

File: experiments/AMIP/user_io/leaderboard/data_sources.jl [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+function Base.close(ds::ObsDataSource)
+    NCDatasets.close(ds.ncdataset)
+end
+
+"""
+
+    A struct to describe some simulation data.
+
+"""
+struct SimDataSource
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/data_sources.jl:68]: Same here, we will need the height information in the future.

File: experiments/AMIP/user_io/leaderboard/data_sources.jl [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    var = get(sim; short_name, reduction, period)
+
+    lonlat = (var.dims["lon"], var.dims["lat"])
+
+    return SimDataSource(path, short_name, reduction, period, var, lonlat)
+end
+
+function data_at_date(sim_ds::SimDataSource, target_date::Dates.DateTime)
+    start_date = Dates.DateTime(sim_ds.var.attributes["start_date"])
+    time_diff_seconds = Dates.value(target_date - start_date) / 1e3
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/data_sources.jl:109]: Why is it divided by 1e3?

File: experiments/AMIP/user_io/leaderboard/utils.jl [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    abs(maximum(lon)) <= 1 && error("longitude should be in degrees")
+    abs(maximum(lat)) <= 1 && error("latitude should be in degrees")
+
+    isequispaced(lon) || error("Longitude is not equispaced")
+    isequispaced(lat) || error("Latitude is not equispaced")
+
+    dlon_rad = deg2rad(abs(lon[begin + 1] - lon[begin]))
+    dlat_rad = deg2rad(abs(lat[begin + 1] - lat[begin]))
+
+    return [sind(90 + lat1) * dlon_rad * dlat_rad for _ in lon, lat1 in lat]
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/utils.jl:56]:
```suggestion
    return [cosd(lat1) * dlon_rad * dlat_rad for _ in lon, lat1 in lat]
```

File: experiments/AMIP/user_io/leaderboard/utils.jl [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+"""
+
+Compute the integration weights for a first-order spherical integration.
+"""
+function integration_weights(lonlat::Tuple{<:AbstractArray, <:AbstractArray})
+    lon, lat = lonlat
+
+    # We are also assuming that they are in degrees
+    abs(maximum(lon)) <= 1 && error("longitude should be in degrees")
+    abs(maximum(lat)) <= 1 && error("latitude should be in degrees")
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/utils.jl:48]: If it's in radian, abs(maximum(lon)) is likely `pi` and abs(maximum(lat)) is likely `pi / 2`, so still larger than 1.

**Review by szy21 — COMMENTED** [review]
Thanks Gabriele! Here are a few minor comments. I need to spend more time understanding how sampling over time works, but other than that it looks good. Also If I understand correctly it is now plotting the rmse map, not the bias map you showed in the comment?

File: experiments/AMIP/user_io/leaderboard/data_sources.jl (hunk 5 above)
Comment (Sbozzolo) [reply experiments/AMIP/user_io/leaderboard/data_sources.jl:109]: Dates.value is in milliseconds, but I changed to make it more readable.

File: experiments/AMIP/user_io/leaderboard/utils.jl (hunk 7 above)
Comment (Sbozzolo) [reply experiments/AMIP/user_io/leaderboard/utils.jl:48]: Indeed.

File: experiments/AMIP/user_io/leaderboard/compare_with_obs.jl (hunk 2 above)
Comment (szy21) [reply experiments/AMIP/user_io/leaderboard/compare_with_obs.jl]: I don't care too much about it. So don't worry about it if it's complicated.

File: experiments/AMIP/user_io/leaderboard.jl
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard.jl:1]: What is this file for?

File: experiments/AMIP/user_io/leaderboard/utils.jl [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+Integrate `data` on a sphere with a first-order scheme. `data` has to be discretized on
+`lonlat`.
+"""
+function integrate_on_sphere(data::AbstractArray, lonlat::Tuple{<:AbstractArray, <:AbstractArray})
+    lon, lat = lonlat
+    size_data = size(data)
+    exp_size = (length(lon), length(lat))
+    size_data == exp_size || error("Inconsistent dimensions $size_data != $exp_size")
+    return sum(data .* integration_weights(lonlat))
```
Comment (szy21) [inline experiments/AMIP/user_io/leaderboard/utils.jl:80]:
```suggestion
    return sum(data .* integration_weights(lonlat)) ./ sum(interpolation_weights(lonlat))
```
Or is the sum of weights 1 here?

**Review by szy21 — APPROVED** [review]
Thank you!

File: experiments/AMIP/user_io/leaderboard/utils.jl (hunk 8 above)
Comment (Sbozzolo) [reply experiments/AMIP/user_io/leaderboard/utils.jl:80]: They are 1 in the limit of infinite resolution. It's a good idea to normalize them

File: experiments/AMIP/user_io/leaderboard.jl
Comment (szy21) [reply experiments/AMIP/user_io/leaderboard.jl:1]: No it's fine. Just curious.

**Pull request merged**